torch/optim/adam.py

```diff
@@ -486,11 +486,14 @@ def _multi_tensor_adam(params: List[Tensor],
             device_grads = torch._foreach_neg(device_grads)
 
         # Handle complex parameters
-        device_grads = [torch.view_as_real(x) if torch.is_complex(x) else x for x in device_grads]
-        device_exp_avgs = [torch.view_as_real(x) if torch.is_complex(x) else x for x in device_exp_avgs]
-        device_exp_avg_sqs = [torch.view_as_real(x) if torch.is_complex(x) else x for x in device_exp_avg_sqs]
-        device_max_exp_avg_sqs = [torch.view_as_real(x) if torch.is_complex(x) else x for x in device_max_exp_avg_sqs]
-        device_params = [torch.view_as_real(x) if torch.is_complex(x) else x for x in device_params]
+        for i in range(len(device_params)):
+            if torch.is_complex(device_params[i]):
+                device_params[i] = torch.view_as_real(device_params[i])
+                device_grads[i] = torch.view_as_real(device_grads[i])
+                device_exp_avgs[i] = torch.view_as_real(device_exp_avgs[i])
+                device_exp_avg_sqs[i] = torch.view_as_real(device_exp_avg_sqs[i])
+                if len(device_max_exp_avg_sqs) > 0:
+                    device_max_exp_avg_sqs[i] = torch.view_as_real(device_max_exp_avg_sqs[i])
```

Comment:
I just noticed something! To uphold previous logic, even though the difference is ever so slight, this should be more like
```suggestion
            if torch.is_complex(device_grads[i]):
                device_grads[i] = torch.view_as_real(device_grads[i])
            if torch.is_complex(device_exp_avgs[i]):
                device_exp_avgs[i] = torch.view_as_real(device_exp_avgs[i])
            if torch.is_complex(device_exp_avg_sqs[i]):
                device_exp_avg_sqs[i] = torch.view_as_real(device_exp_avg_sqs[i])
            if len(device_max_exp_avg_sqs) > 0 and torch.is_complex(device_max_exp_avg_sqs[i]):
                device_max_exp_avg_sqs[i] = torch.view_as_real(device_max_exp_avg_sqs[i])
```

Reply:
This shouldn't change compile time, if I understand correctly, but I would be curious about the numbers after this.

Reply:
>  To uphold previous logic, even though the difference is ever so slight, this should be more like

This is never required, because there is a 1-1 correspondance between param and `grads`, `exp_avgs`, `exp_avg_sqs` etc. except for `max_exp_avg_sqs` which is only initialized to len > 1 for 'amsdelta' flag for the entire group. See the `single_tensor_{}` impls for confirmation, it always shortcuts the `is_complex` checks.

Reply:
> This shouldn't change compile time, if I understand correctly, but I would be curious about the numbers after this.

I believe it would actually, as the long compile time is predicated on calling the check on each of the lists instead of just one list.

Reply:
Ah I see, seeing that the single_tensor had made the assumption already and that this would remove a lot of the perf wins, it should be safe to keep making this assumption.

Reply:
Ok, if all is good, I'll merge this one. Thanks for the help!